### PR TITLE
[FIX]  Wallet connect signed typed and eth sign throwing error back to the dapp

### PR DESCRIPTION
**Pull request by tommasini**
**Description**
Wallet connect default sign typed data function we redirect it to v3 and if the dapp send a message with array structured we do not support it. 
We also were not returning back the eth sign function if it went wrong back to the dapp.

**Proposed Solution**
When a user signs a message that is only supported with v4 with the default sign method of wallet connect, we throw the error back to the dapp and say to our user that the sign failed.
We also send back to the dapp the error if eth sign fails.

**Screenshots/Recordings**
https://recordit.co/mH3Es44m1J

**Test Cases**
Case1:
* Clone metamask test dapp from this github https://github.com/tommasini/test-dapp
* switch to branch `wc-support`
* Open the project and on terminal type `yarn setup` and then `yarn start`
* Tested all the functions to test with wallet connect at the bottom of test dapp

Case2:
* Use wallet connect example https://example.walletconnect.org/
* Tested as well all the functions there

**Code impact**
Low

**Issue**

Progresses #https://github.com/MetaMask/metamask-mobile/issues/4441, https://github.com/MetaMask/metamask-mobile/issues/3887, https://github.com/MetaMask/metamask-mobile/issues/3580

**Checklist**

* [ ] There is a related GitHub issue
* [ ] Tests are included if applicable
* [ ] Any added code is fully documented

### CODE REVIEW

**Comment by kevinwo** [comment]
While it looks like this PR resolves the silent failure when attempting to sign with V4, will there be a follow-on PR then that upgrades the WalletConnect library to >= 1.8.0 and support V4 signing?

**Comment by tommasini** [comment]
Hello @kevinwo , thanks for testing the PR!

I'm not sure what you are asking but you can do a sign request V4 if you stablish connection with wallet connect and send a custom request with the method `eth_signTypedData_v4`, just the default sign method of wallet connect will be redirected to the V3

**Comment by kevinwo** [comment]
> you can do a sign request V4 if you stablish connection with wallet connect and send a custom request with the method `eth_signTypedData_v4`

@tommasini Ahh yes, that makes sense. Thank you!

File: app/components/UI/TypedSign/index.js [hunk 1]
```diff
@@ -135,15 +141,29 @@ class TypedSign extends PureComponent {
...
+    }
+
+    if (!messageError) {
+      TypedMessageManager.setMessageStatusSigned(messageId, rawSig);
+      this.showWalletConnectNotification(messageParams, true);
+    } else {
+      TypedMessageManager.setMessageStatusSigned(messageId, messageError);
+
+      this.showWalletConnectNotification(messageParams, false, true);
+    }
```
Comment (gantunesr) [inline app/components/UI/TypedSign/index.js:166]: Is there a reason why this logic can't be inside the `try-catch`?

**Review by gantunesr — COMMENTED** [review]
LGTM. Just a couple of small comments

File: app/components/UI/TypedSign/index.js (hunk 1 above)
Comment (tommasini) [reply app/components/UI/TypedSign/index.js:166]: Don't make any sense to be outside of the try-catch, amazing catch! Thanks for the review

File: app/core/RPCMethods/RPCMethodMiddleware.ts [hunk 2]
```diff
@@ -313,6 +313,7 @@ export const getRpcMethodMiddleware = ({
 
           res.result = rawSig;
         } else {
+          res.result = 'eth_sign requires 32 byte message hash';
           throw ethErrors.rpc.invalidParams(
             'eth_sign requires 32 byte message hash',
```
[the remark this answers is not in the document]
Comment (tommasini) [reply app/core/RPCMethods/RPCMethodMiddleware.ts]: Done

**Review by gantunesr — APPROVED** [review]
LGTM

**Review by cortisiko — CHANGES_REQUESTED** [review]
@tommasini swinging this back your way! 


**Issue 1**
On your clone repo when I tried connecting to wallet connect by scanning the QR code nothing happens. I am not prompted with the connect modal. Furthermore, when I attempt to scan the QR code again, I get prompted that the QR was already scanned. See [recording](http://recordit.co/QXyqqUVtxP ).


**Issue 2**
When I tried signing WC v4 I get Error: `TypedMessageManager: Message not found for id: undefined. `

see [recording](http://recordit.co/DFJwWe0nQ1).  

To reproduce:
Connect to wallet connect and try signing WC sign Typed Data V4

**Issue 3**
Eth sign is broken on your repo. I get `Error: eth_sign requires 32 byte message hash`

To reproduce:
Connect to the test dapp on your repo
Tap Eth Personal Sign
Notice the error

See [recording](http://recordit.co/19nVLbETLK)

Keep in mind, the regular test dapp is working as expected!

**Comment by tommasini** [comment]
Thank you for testing @cortisiko!

Issue 1:
I'm trying to reproduce but every time I press connect again after I disconnect, everything went normal 
Issue 2:
I'm not being able to reproduce it, and I tried via in-app browser, desktop and chrome mobile
Issue 3:
 It's the expected behaviour, it's to demonstrate that we are sending the error back to the dapp (we can confirm seeing the response on the test dapp)

**Review by cortisiko — APPROVED** [review]
@tommasini thanks for the update. This PR is QA passed. Things I tested:
- signing with WC on a couple dapps such as opensea and rariable 
- signing on our test dapp
- signing with WC on your (@tommasini) test dapp
- signing WC with keystone
- signing regular dapps with keystone
- WC transactions

**Pull request merged**